fdb-extensions/src/main/java/com/apple/foundationdb/async/RTree.java

```diff
@@ -0,0 +1,2571 @@
+            Preconditions.checkArgument(slot instanceof ChildSlot);
+            children.add(slotIndex, (ChildSlot)slot);
+        }
+
+        @Nonnull
+        @Override
+        public IntermediateNode newOfSameKind(@Nonnull final byte[] nodeId) {
+            return new IntermediateNode(nodeId, Lists.newArrayList());
+        }
+
+        @Nonnull
+        public String getPlotMbrs() {
+            return getSlots().stream()
+                    .map(child -> child.getMbr().toPlotString())
+                    .collect(Collectors.joining("\n"));
+        }
+    }
+
+    /**
+     * Abstract base class for all node slots. Holds a Hilbert value and a key. The semantics of these attributes
+     * is refined in the subclasses {@link ItemSlot} and {@link ChildSlot}.
+     */
+    private abstract static class NodeSlot {
+        @Nonnull
+        protected BigInteger hilbertValue;
+
+        @Nonnull
+        protected Tuple key;
+
+        protected NodeSlot(@Nonnull final BigInteger hilbertValue, @Nonnull final Tuple key) {
+            this.hilbertValue = hilbertValue;
+            this.key = key;
+        }
+
+        @Nonnull
+        public BigInteger getHilbertValue() {
+            return hilbertValue;
+        }
+
+        @Nonnull
+        public Tuple getKey() {
+            return key;
+        }
+
+        /**
+         * Create a tuple for the key part of this slot. This tuple is used when the slot is persisted in the database.
+         * Note that the serialization format is not yet finalized.
+         * @return a new tuple
+         */
+        @Nonnull
+        protected abstract Tuple getSlotKey();
+
+        /**
+         * Create a tuple for the value part of this slot. This tuple is used when the slot is persisted in the database.
+         * Note that the serialization format is not yet finalized.
+         * @return a new tuple
+         */
+        @Nonnull
+        protected abstract Tuple getSlotValue();
+    }
+
+    /**
+     * An item slot that is used by {@link LeafNode}s. Holds the actual data of the item as well as the items Hilbert
+     * value and its key.
+     */
+    static class ItemSlot extends NodeSlot {
+        @Nonnull
+        private final Tuple value;
+        @Nonnull
+        private final Point position;
+
+        public ItemSlot(@Nonnull final BigInteger hilbertValue, @Nonnull final Tuple key, @Nonnull final Tuple value,
+                        @Nonnull final Point position) {
+            super(hilbertValue, key);
+            this.value = value;
+            this.position = position;
+        }
+
+        @Nonnull
+        public Tuple getValue() {
+            return value;
+        }
+
+        @Nonnull
+        public Point getPosition() {
+            return position;
+        }
+
+        @Nonnull
+        @Override
+        protected Tuple getSlotKey() {
+            return Tuple.from((byte)0x00, getHilbertValue(), getKey());
```

Comment:
This puts the key in a nested tuple. That's fine, though I think you could also avoid the nested tuple by making this `Tuple.from(0x00, getHilbertValue()).addAll(getKey())` (effectively concatenating the tuples). In terms of semantics, I think those are all the same, but it may save a few bytes

Reply:
the Hilbert value is not a `Tuple` but just a `BigInteger`.

Reply:
I was talking about `getKey()`